pallets/contracts2/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+            )
+        })
+    }
+
+    /// Enriches `result` of executing a smart contract with actual weight,
+    /// accounting for the consumed gas.
+    fn handle_error<A>(
+        base_weight: Weight,
+        result: ContractResult<Result<A, DispatchError>>,
+    ) -> DispatchResultWithPostInfo {
+        let post_info = Some(result.gas_consumed.saturating_add(base_weight)).into();
+        match result.result {
+            Ok(_) => Ok(post_info),
+            Err(error) => Err(DispatchErrorWithPostInfo { post_info, error }),
+        }
+    }
+}
+
+/// The `Call` enums of various pallets that the contracts pallet wants to know about.
+pub enum CommonCall<T>
+where
+    T: Config + pallet_asset::Config,
+{
+    Asset(pallet_asset::Call<T>),
+    Contracts(Call<T>),
+}
+
+/// An encoding of `func_id` into the encoding `0x_S_P_E_V`,
+/// with each letter being a byte.
+struct FuncId {
+    /// Decides the version of the `extrinsic`.
+    version: u8,
+    /// Decides the `extrinsic` within the `pallet`.
+    /// This isn't necessarily the same as the index within the pallet.
+    extrinsic: u8,
+    /// Decides the `pallet` within the runtime.
+    /// This isn't necessarily the same as the index within the runtime.
+    pallet: u8,
+    /// Decides the scheme to use when interpreting `(extrinsic, pallet, version)`.
+    scheme: u8,
+}
+
+/// Splits the `func_id` given from a smart contract into
+/// the encoding `(scheme: u8, pallet: u8, extrinsic: u8, version: u8)`.
+fn split_func_id(func_id: u32) -> FuncId {
+    let extract = |which| (func_id >> which * 8) as u8;
+    FuncId {
+        version: extract(0),
+        extrinsic: extract(1),
+        pallet: extract(2),
+        scheme: extract(3),
+    }
+}
+
+/// Set current DID to that of `key` and returns the old DID.
+fn set_current_did_to_key<T: Config>(key: &T::AccountId) -> Option<IdentityId> {
+    let old_did = Context::current_identity::<Identity<T>>();
+    let caller_did = Identity::<T>::key_to_identity_dids(key);
+    Context::set_current_identity::<Identity<T>>(Some(caller_did));
+    old_did
+}
+
+fn decode<V: Decode, T: Config>(input: &mut &[u8]) -> Result<V, DispatchError> {
+    <_>::decode(input).map_err(|_| pallet_contracts::Error::<T>::DecodingFailed.into())
+}
+
+fn construct_call<T>(func_id: FuncId, input: &mut &[u8]) -> Result<CommonCall<T>, DispatchError>
+where
+    T: Config + pallet_asset::Config,
+{
+    /// Decode type from `input`.
+    macro_rules! decode {
+        () => {
+            decode::<_, T>(input)?
+        };
+    }
+
+    /// Pattern match on functions `0x00_pp_ee_00`.
+    macro_rules! on {
+        ($p:pat, $e:pat) => {
+            FuncId {
+                scheme: 0,
+                pallet: $p,
+                extrinsic: $e,
+                version: 0,
+            }
+        };
+    }
+
+    // TODO(Centril): Should we charge weight depending on branch?
+    // That would require us to benchmark the decoding of each and every call.
```

Comment:
Thoughts?

Reply:
I think charging weights here would make the logic too complex.

It would be better to just apply sane limits on the length of `input` based on `func_id`.